### PR TITLE
sponsored project model updates

### PR DESCRIPTION
**Pull request by romanetar**
ref https://tipit.avaza.com/project/view/188886#!tab=task-pane&groupby=ProjectSection&view=vertical&task=2979686&fileview=grid

ref https://tipit.avaza.com/task#!sortby=DateUpdatedDesc&task=2980639

Signed-off-by: romanetar <roman_ag@hotmail.com>

### CODE REVIEW

File: app/Models/Foundation/Main/Companies/SponsoredProject.php [hunk 1]
```diff
@@ -107,6 +144,136 @@ public function getSlug(): string
...
+     */
+    public function setShouldShowOnNavBar(bool $should_show_on_nav_bar): void
+    {
+        $this->should_show_on_nav_bar = $should_show_on_nav_bar;
+    }
+
+    /**
+     * @return string
+     */
+    public function getLearnMoreLink(): string
```
Comment (smarcet) [inline app/Models/Foundation/Main/Companies/SponsoredProject.php:182]: this should return nullable (?string)

File: app/Models/Foundation/Main/Companies/SponsoredProject.php [hunk 2]
```diff
@@ -107,6 +144,136 @@ public function getSlug(): string
...
+     */
+    public function setLearnMoreLink(string $learn_more_link): void
+    {
+        $this->learn_more_link = $learn_more_link;
+    }
+
+    /**
+     * @return string
+     */
+    public function getLearnMoreText(): string
```
Comment (smarcet) [inline app/Models/Foundation/Main/Companies/SponsoredProject.php:198]: this should return nullable (?string)

File: tests/OAuth2SponsoredProjectsApiTest.php [hunk 3]
```diff
@@ -67,8 +67,8 @@ public function testGelAll()
         );
 
         $content = $response->getContent();
-        $companies = json_decode($content);
-        $this->assertTrue(!is_null($companies));
+        $sponsored_projects = json_decode($content);
```
Comment (smarcet) [inline tests/OAuth2SponsoredProjectsApiTest.php:70]: @romanetar we need to update the add and update test to add and check the new attributes

**Review by smarcet — CHANGES_REQUESTED** [review]
@romanetar please check comments

File: app/Models/Foundation/Main/Companies/SponsoredProject.php (hunk 2 above)
Comment (romanetar) [reply app/Models/Foundation/Main/Companies/SponsoredProject.php]: done

File: app/Models/Foundation/Main/Companies/SponsoredProject.php (hunk 1 above)
Comment (romanetar) [reply app/Models/Foundation/Main/Companies/SponsoredProject.php]: done

File: tests/OAuth2SponsoredProjectsApiTest.php (hunk 3 above)
Comment (romanetar) [reply tests/OAuth2SponsoredProjectsApiTest.php:123]: done

**Comment by romanetar** [comment]
@smarcet everything already fixed as requested

File: app/Http/Controllers/Apis/Protected/Main/Factories/SponsoredProjectValidationRulesFactory.php [hunk 4]
```diff
@@ -30,12 +30,22 @@ public static function build(array $data, $update = false){
...
         }
         return [
             'name' => 'required|string',
             'description' => 'sometimes|string',
             'is_active' => 'sometimes|boolean',
+            'nav_bar_title' => 'sometimes|string',
+            'should_show_on_nav_bar' => 'sometimes|boolean',
+            'learn_more_link' => 'sometimes|string',
+            'learn_more_text' => 'sometimes|string',
+            'site_url' => 'sometimes|string',
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Main/Factories/SponsoredProjectValidationRulesFactory.php:48]: should be https://laravel.com/docs/9.x/validation#rule-url

File: app/Http/Controllers/Apis/Protected/Main/Factories/SponsoredProjectValidationRulesFactory.php [hunk 5]
```diff
@@ -30,12 +30,22 @@ public static function build(array $data, $update = false){
...
+                'site_url' => 'sometimes|string',
             ];
         }
         return [
             'name' => 'required|string',
             'description' => 'sometimes|string',
             'is_active' => 'sometimes|boolean',
+            'nav_bar_title' => 'sometimes|string',
+            'should_show_on_nav_bar' => 'sometimes|boolean',
+            'learn_more_link' => 'sometimes|string',
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Main/Factories/SponsoredProjectValidationRulesFactory.php:46]: should be https://laravel.com/docs/9.x/validation#rule-url

File: app/Http/Controllers/Apis/Protected/Main/Factories/SponsoredProjectValidationRulesFactory.php [hunk 6]
```diff
@@ -30,12 +30,22 @@ public static function build(array $data, $update = false){
                 'name' => 'sometimes|string',
                 'description' => 'sometimes|string',
                 'is_active' => 'sometimes|boolean',
+                'nav_bar_title' => 'sometimes|string',
+                'should_show_on_nav_bar' => 'sometimes|boolean',
+                'learn_more_link' => 'sometimes|string',
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Main/Factories/SponsoredProjectValidationRulesFactory.php:35]: should be https://laravel.com/docs/9.x/validation#rule-url

File: app/Http/Controllers/Apis/Protected/Main/Factories/SponsoredProjectValidationRulesFactory.php [hunk 7]
```diff
@@ -30,12 +30,22 @@ public static function build(array $data, $update = false){
...
+                'learn_more_text' => 'sometimes|string',
+                'site_url' => 'sometimes|string',
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Main/Factories/SponsoredProjectValidationRulesFactory.php:37]: should be https://laravel.com/docs/9.x/validation#rule-url

**Review by smarcet — CHANGES_REQUESTED** [review]
@romanetar please review comments

File: app/Http/Controllers/Apis/Protected/Main/Factories/SponsoredProjectValidationRulesFactory.php (hunk 6 above)
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Main/Factories/SponsoredProjectValidationRulesFactory.php]: done

File: app/Http/Controllers/Apis/Protected/Main/Factories/SponsoredProjectValidationRulesFactory.php (hunk 5 above)
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Main/Factories/SponsoredProjectValidationRulesFactory.php]: done

File: app/Http/Controllers/Apis/Protected/Main/Factories/SponsoredProjectValidationRulesFactory.php (hunk 4 above)
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Main/Factories/SponsoredProjectValidationRulesFactory.php]: done

File: app/ModelSerializers/Companies/SponsoredProjectSerializer.php [hunk 8]
```diff
@@ -23,6 +23,12 @@ final class SponsoredProjectSerializer extends SilverStripeSerializer
         'Description' => 'description:json_string',
         'Slug' => 'slug:json_string',
         'Active' => 'is_active:json_boolean',
+        'NavBarTitle' => 'nav_bar_title:json_string',
+        'ShouldShowOnNavBar' => 'should_show_on_nav_bar:json_boolean',
+        'LearnMoreLink' => 'learn_more_link:json_string',
+        'LearnMoreText' => 'learn_more_text:json_string',
+        'SiteURL' => 'site_url:json_string',
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/ModelSerializers/Companies/SponsoredProjectSerializer.php]: done

File: app/Http/Controllers/Apis/Protected/Main/OAuth2SponsoredProjectApiController.php [hunk 9]
```diff
@@ -423,4 +426,34 @@ public function getSupportingCompany($id, $sponsorship_type_id, $company_id){
...
+     * @param LaravelRequest $request
+     * @param $project_id
+     * @return mixed
+     */
+    public function addSponsoredProjectLogo(LaravelRequest $request, $project_id)
+    {
+        return $this->processRequest(function () use ($request, $project_id) {
+            $file = $request->file('file');
+            if (is_null($file)) {
+                return $this->error412(array('file param not set!'));
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Main/OAuth2SponsoredProjectApiController.php:440]: this is not needed ( dont enclose on array ) error412 already does

File: app/Http/Controllers/Apis/Protected/Main/OAuth2SponsoredProjectApiController.php [hunk 10]
```diff
@@ -423,4 +426,34 @@ public function getSupportingCompany($id, $sponsorship_type_id, $company_id){
...
+        });
+    }
+
+    /**
+     * @param $project_id
+     * @return \Illuminate\Http\JsonResponse|mixed
+     */
+    public function deleteSponsoredProjectLogo($project_id){
+        return $this->processRequest(function () use ($project_id) {
+            $this->service->deleteLogo($project_id);
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Main/OAuth2SponsoredProjectApiController.php:454]: please do parseInt here

File: app/Http/Controllers/Apis/Protected/Main/OAuth2SponsoredProjectApiController.php [hunk 11]
```diff
@@ -423,4 +426,34 @@ public function getSupportingCompany($id, $sponsorship_type_id, $company_id){
...
+     * @return mixed
+     */
+    public function addSponsoredProjectLogo(LaravelRequest $request, $project_id)
+    {
+        return $this->processRequest(function () use ($request, $project_id) {
+            $file = $request->file('file');
+            if (is_null($file)) {
+                return $this->error412(array('file param not set!'));
+            }
+            $logo = $this->service->addLogo($project_id, $file);
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Main/OAuth2SponsoredProjectApiController.php:442]: do parseInt here

File: database/seeders/ApiEndpointsSeeder.php [hunk 12]
```diff
@@ -6823,6 +6823,30 @@ private function seedSponsoredProjectsEndpoints(){
                         sprintf(SponsoredProjectScope::Read, $current_realm)
                     ]
                 ],
+                [
+                    'name' => 'add-sponsored-project-logo',
+                    'route' => '/api/v1/sponsored-projects/{id}/logo',
+                    'http_method' => 'POST',
+                    'scopes' => [
+                        sprintf(CompanyScopes::Write, $current_realm)
```
Comment (smarcet) [inline database/seeders/ApiEndpointsSeeder.php:6831]: this is not the correct one
u should use SponsoredProjectScope::Write

File: database/seeders/ApiEndpointsSeeder.php [hunk 13]
```diff
@@ -6823,6 +6823,30 @@ private function seedSponsoredProjectsEndpoints(){
...
+                        IGroup::SuperAdmins,
+                        IGroup::Administrators,
+                    ]
+                ],
+                [
+                    'name' => 'delete-sponsored-project-logo',
+                    'route' => '/api/v1/sponsored-projects/{id}/logo',
+                    'http_method' => 'DELETE',
+                    'scopes' => [
+                        sprintf(CompanyScopes::Write, $current_realm)
```
Comment (smarcet) [inline database/seeders/ApiEndpointsSeeder.php:6843]: same here

File: app/Http/Controllers/Apis/Protected/Main/OAuth2SponsoredProjectApiController.php (hunk 9 above)
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Main/OAuth2SponsoredProjectApiController.php:440]: done

File: app/Http/Controllers/Apis/Protected/Main/OAuth2SponsoredProjectApiController.php (hunk 10 above)
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Main/OAuth2SponsoredProjectApiController.php:454]: done

File: app/Http/Controllers/Apis/Protected/Main/OAuth2SponsoredProjectApiController.php (hunk 11 above)
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Main/OAuth2SponsoredProjectApiController.php:442]: done

File: database/seeders/ApiEndpointsSeeder.php (hunk 12 above)
Comment (romanetar) [reply database/seeders/ApiEndpointsSeeder.php:6831]: done

File: database/seeders/ApiEndpointsSeeder.php (hunk 13 above)
Comment (romanetar) [reply database/seeders/ApiEndpointsSeeder.php:6843]: done

**Review by smarcet — APPROVED** [review]
LGTM

**Pull request merged**